diagnostics/monitoring_server.rs

```diff
@@ -0,0 +1,90 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use std::{
+    convert::Infallible,
+    net::SocketAddr,
+    sync::{Arc, Mutex},
+    thread,
+};
+
+use futures::executor::block_on;
+use hyper::{
+    header::{CONNECTION, CONTENT_LENGTH, CONTENT_TYPE},
+    service::{make_service_fn, service_fn},
+    Body, Request, Response, Server, StatusCode,
+};
+use logger::{debug, trace};
+use serde_json::json;
+use tokio::{runtime::Runtime, task};
+
+use crate::Diagnostics;
+
+#[derive(Debug)]
+pub struct MonitoringServer {
+    diagnostics: Arc<Diagnostics>,
+    port: u16,
+    is_enabled: bool,
+}
+
+impl MonitoringServer {
+    pub fn new(diagnostics: Arc<Diagnostics>, port: u16, is_enabled: bool) -> Self {
+        Self { diagnostics, port, is_enabled }
+    }
+
+    pub fn may_start(&self) {
+        if !self.is_enabled {
+            return;
+        }
+
+        let addr = SocketAddr::from(([0, 0, 0, 0], self.port));
+        let diagnostics = self.diagnostics.clone();
+
+        task::spawn(async move {
```

Comment:
A `tokio` task for the created `hyper` service.

Reply:
A little weird, but it's probably worth defensively marking `may_start` as `async` since it requires being run in a `tokio` runtime. That will force `Server::open` to be `async` as well.

Reply:
Which means that I'll move task::spawn for the monitoring start to `Server::open`, which is now:
```
        diagnostics_manager.may_start_monitoring();
        diagnostics_manager.may_start_reporting();
```
And we'll spawn it even if monitoring is not needed (if it returns after `!self.is_enabled`). Is it fine?

Reply:
But `Server::open` is not async, how can I `.await` there without a runtime?

Reply:
Done!